### PR TITLE
updates eip5573 to reflect changes in CACAO and UCAN specs

### PR DESCRIPTION
**Pull request by chunningham**
This PR updates EIP5573 to enable easier integration with CACAO and UCAN by:
- replacing the `tar` and `ext` fields with `att` and `prf`
  - `att` containing URIs as keys and objects as values
    - each of these objects contains `<namespace>/<name>` ability strings as keys and an array of note-bene/restriction objects as values
      - ability strings are restricted to alphanumeric characters + `.+*-_`
    - note-bene objects can contain any JSON
  - `prf` containing an array of CIDs representing prior authorisations
- requiring maps within the `recap-object` to have lexicographically ordered keys
- updating the statement generation to segregate by resource and namespace
- requiring that there be only one `recap-uri` per SIWE message that must appear as the final entry in the resource list
- removing the namespace from the `recap-uri`

By matching the `recap-object` to the `att` and `prf` structure of ucan/cacao, interoperability between them becomes possible by treating a SIWE-recap as just one way of representing the same information, for signature verification purposes.

### CODE REVIEW

**Review by awoie — APPROVED** [review]
lgtm

File: EIPS/eip-5573.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        "format": "CID"
+      },
+      "minItems": 1
+    }
+  }
+}
+```
+
+A ReCap Details Object defines the following properties:
+- `att`: (CONDITIONAL) If present, `att` MUST be a JSON object where each key is a URI and each value is an object containing namespaced-actions as keys and a corrosponding value which is an array of qualifications to the action (i.e. a restriction or requirement). The keys of the object MUST be ordered lexicographically.
```
Comment (cobward) [inline EIPS/eip-5573.md:139]:
```suggestion
- `att`: (CONDITIONAL) If present, `att` MUST be a JSON object where each key is a URI and each value is an object containing Ability Strings as keys and a corresponding value which is an array of qualifications to the action (i.e. a restriction or requirement). The keys of the object MUST be ordered lexicographically.
```

File: EIPS/eip-5573.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+- `prf`: (CONDITIONAL) If present, `prf` MUST be a JSON array of string values with at least one entry where each value is a valid Base58-encoded CID which identifies a parent capability, authorising one or more of the entries in `att` if the SIWE `address` does not identify the controller of the `att` entries.
```
Comment (cobward) [inline EIPS/eip-5573.md:140]: Make it clear that a parent capability should authorize the Ethereum account that is signing this ReCap. Something like:
```suggestion
- `prf`: (CONDITIONAL) If present, `prf` MUST be a JSON array of string values with at least one entry where each value is a valid Base58-encoded CID which identifies a parent capability, authorising the Ethereum account for one or more of the entries in `att` if the SIWE `address` does not identify the controller of the `att` entries.
```

Comment (cobward) [inline EIPS/eip-5573.md:140]: Use American spelling of authorize
```suggestion
- `prf`: (CONDITIONAL) If present, `prf` MUST be a JSON array of string values with at least one entry where each value is a valid Base58-encoded CID which identifies a parent capability, authorizing one or more of the entries in `att` if the SIWE `address` does not identify the controller of the `att` entries.
```

File: EIPS/eip-5573.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+Objects in the `att` field (including nested objects) MUST have their keys ordered lexicographically with two steps:
+1. Sort by byte value.
+2. If a string starts with another, the shorter string comes first (e.g. `msg/send` comes before `msg/send-to`)
+
+This is the same as the `Array.sort()` method in Javascript. In the example below, `crud/delete` must appear before `crud/update` and `other/action`, similarly `msg/recieve` must appear before `msg/send`.
```
Comment (cobward) [inline EIPS/eip-5573.md:146]:
```suggestion
This is the same as the `Array.sort()` method in Javascript. In the example below, `crud/delete` must appear before `crud/update` and `other/action`, similarly `msg/receive` must appear before `msg/send`.
```

File: EIPS/eip-5573.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+         "crud/delete": [],
+         "crud/update": [],
+         "other/action": []
+      },
+      "mailto:username@example.com":{
+          "msg/receive": [{
+              "max_count": 5,
+              "templates": ["newsletter", "marketing"]
+          }],
+          "msg/send": [{ "to": "someone@email.com" }, { "to": "joe@email.com" }],
```
Comment (cobward) [inline EIPS/eip-5573.md:162]:
```suggestion
          "msg/send": [{ "to": "someone@email.com" }, { "to": "joe@email.com" }]
```

File: EIPS/eip-5573.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+This example is expected to transform into the following `recap-transformed-statement` (for `URI` of `https://example.com`):
+
+```text
+I further authorize did:key:example to perform the following actions on my behalf: (1) "crud": "delete", "update" for "https://example.com/pictures". (2) "other": "action" for "https://example.com/pictures". (3) "msg": "recieve", "send" for "mailto:username@example.com".
+```
+
+This example is also expected to transform into the following `recap-uri`:
+
+```text
+urn:recap:eyJhdHQiOnsiaHR0cHM6Ly9leGFtcGxlLmNvbS9waWN0dXJlcyI6eyJjcnVkL2RlbGV0ZSI6W10sImNydWQvdXBkYXRlIjpbXSwib3RoZXIvYWN0aW9uIjpbXX0sIm1haWx0bzp1c2VybmFtZUBleGFtcGxlLmNvbSI6eyJtc2cvcmVjaWV2ZSI6W3sibWF4X2NvdW50Ijo1LCJ0ZW1wbGF0ZXMiOlsibmV3c2xldHRlciIsIm1hcmtldGluZyJdfV0sIm1zZy9zZW5kIjpbeyJ0byI6ImpvZUBlbWFpbC5jb20ifV19fSwicHJmIjpbImJhZnliZWlnazdseTNwb2c2dXVweGt1M2I2YnViaXJyNDM0aWI2dGZheW12b3g2Z290YWFhYWFhYWFhIl19
```
Comment (cobward) [inline EIPS/eip-5573.md:182]: URI is out of date. Not sure about whether we want padding.
```suggestion
urn:recap:eyJhdHQiOnsiaHR0cHM6Ly9leGFtcGxlLmNvbS9waWN0dXJlcy8iOnsiY3J1ZC9kZWxldGUiOltdLCJjcnVkL3VwZGF0ZSI6W10sIm90aGVyL2FjdGlvbiI6W119LCJtYWlsdG86dXNlcm5hbWVAZXhhbXBsZS5jb20iOnsibXNnL3JlY2VpdmUiOlt7Im1heF9jb3VudCI6NSwidGVtcGxhdGVzIjpbIm5ld3NsZXR0ZXIiLCJtYXJrZXRpbmciXX1dLCJtc2cvc2VuZCI6W3sidG8iOiJzb21lb25lQGVtYWlsLmNvbSJ9LHsidG8iOiJqb2VAZW1haWwuY29tIn1dfX0sInByZiI6WyJiYWZ5YmVpZ2s3bHkzcG9nNnV1cHhrdTNiNmJ1YmlycjQzNGliNnRmYXltdm94NmdvdGFhYWFhYWFhYSJdfQ==
```

File: EIPS/eip-5573.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+```
+
+#### ReCap Translation Algorithm
+
+After applying the ReCap Translation Algorithm on a given SIWE message that MAY include a pre-defined `statement`, the `recap-transformed-statement` in a ReCap SIWE message MUST conform to the following ABNF:
+```text
+recap-transformed-statement = statement recap-preamble 1*(" " recap-statement-entry ".")
+   ; see EIP-4361 for definition of input-statement
+recap-preamble = "I further authorize " uri " to perform the following actions on my behalf:"
```
Comment (cobward) [inline EIPS/eip-5573.md:191]: As above: are we updating the statement to `stated URI` instead of the in-line URI?

File: EIPS/eip-5573.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    - Append the value of the `statement` field of `siwe` to `recap-transformed-statement`.
+    - Append a single space character `" "` to `recap-transformed-statement`.
+- Append the following string to `recap-transformed-statement`: "I further authorize ".
+- Append `uri` to `recap-transformed-statement`.
+- Append the following string to `recap-transformed-statement`: " to perform the following actions on my behalf:".
+- Let `numbering` be an integer starting with 1.
+- Let `attenuations` be the `att` field of the ReCap Details Object
+- For each key and value pair in `attenuations` (starting with the first entry), perform the following:
+  - Let `resource` be the key and `abilities` be the value
+  - Group the keys of the `ability` object by their `ability-namespace`
```
Comment (cobward) [inline EIPS/eip-5573.md:223]:
```suggestion
  - Group the keys of the `abilities` object by their `ability-namespace`
```

File: EIPS/eip-5573.md [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+- Let `uri` be the uri field of the input SIWE message conforming to EIP-4361.
+- Let `recap-uri` be a ReCap URI, which represents the ReCap Capabilities that are to be encoded in the SIWE message, and which contains a ReCap Details Object which conforms to the ReCap Details Object Schema.
+- [Optional] Let `statement` be the statement field of the input SIWE message conforming to EIP-4361.
+Algorithm:
+- Let `recap-transformed-statement` be an empty string value.
+- If `statement` is present, do the following: 
+    - Append the value of the `statement` field of `siwe` to `recap-transformed-statement`.
+    - Append a single space character `" "` to `recap-transformed-statement`.
+- Append the following string to `recap-transformed-statement`: "I further authorize ".
+- Append `uri` to `recap-transformed-statement`.
```
Comment (cobward) [inline EIPS/eip-5573.md:217]: As above: are we updating the statement to `stated URI` instead of the in-line URI?

File: EIPS/eip-5573.md [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+- For each key and value pair in `attenuations` (starting with the first entry), perform the following:
+  - Let `resource` be the key and `abilities` be the value
+  - Group the keys of the `ability` object by their `ability-namespace`
+  - For each `ability-namespace`, perform the following:
+    - Append the string concatenation of `" ("`, `numbering`, `")"` to `recap-transformed-statement`.
+    - Append the string concatenation of `"`, `ability-namespace`, `": ` to `recap-transformed-statement`.
+    - For each `ability-name` in the `ability-namespace` group, perform the following:
+      - Append the string concatenation of `"`, `ability-name`, `"` to `recap-transformed-statement`
+      - If not the final `ability-name`, append `, ` to `recap-transformed-statement`
+    - Append `for "`, `resource`, `".` to `recap-transformed-statement`
```
Comment (cobward) [inline EIPS/eip-5573.md:230]:
```suggestion
    - Append ` for "`, `resource`, `".` to `recap-transformed-statement` 
```

File: EIPS/eip-5573.md [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+- The `Issued At` field MUST be used to specify the issuance date of the ReCap Capabilities.
+- If present, the `Expiration Time` field MUST be used as the expiration time of the ReCap Capabilities, i.e. the time at which the RS will no longer accept an invocation of the capabilities expressed in this form.
+- If present, the `Not Before` field MUST be used as the time that has to expire before the RS starts accepting invocations of the capabilities expressed in the message.
+
+The following is a non-normative example of a SIWE message with the SIWE ReCap Extension:
+```text
+example.com wants you to sign in with your Ethereum account:
+0x0000000000000000000000000000000000000000
+
+I further authorize did:key:example to perform the following actions on my behalf: (1) "example": "append", "read" for "https://example.com". (2) "other": "action" for "https://example.com". (3) "example": "append", "delete" for "my:resource:uri.1". (4) "example": "append" for "my:resource:uri.2". (5) "example": "append" for "my:resource:uri.3".
```
[the remark this answers is not in the document]
Comment (chunningham) [reply EIPS/eip-5573.md]: I think so

File: EIPS/eip-5573.md (hunk 5 above)
Comment (chunningham) [reply EIPS/eip-5573.md]: without padding imo but not a strong opinion

File: EIPS/eip-5573.md [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+## Motivation
+
+SIWE ReCaps unlock integration of protocols and/or APIs for developers by reducing user friction, onchain state and increasing security by introducing informed consent and deterministic capability objects on top of Sign-In With Ethereum (EIP-4361).
+
+While SIWE focuses on authenticating the Ethereum account against the service (relying party or SIWE client) initiating the SIWE flow, there is no canonical way to interact with a third-party service (resource service) on behalf of the authenticated Ethereum account. For example, a relying party might want to interact with another service on behalf of the Ethereum account, for example a service that provides data storage for the Ethereum account. This specification introduces a mechanism, that allows the service (or more generally a delegee) to combine authentication and authorization of such while preserving security and optimizing UX.
+
+Note, this approach is a similar mechanism to combining OpenID Connect (SIWE auth) and OAuth2 (SIWE ReCap) whereas SIWE ReCap follows an Object Capability-based approach.
```
[the remark this answers is not in the document]
Comment (chunningham) [reply EIPS/eip-5573.md]: Also not sure of the intended meaning here, tried to clarify by relating siwe to IODC and recap to OAuth2

File: EIPS/eip-5573.md [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+## Motivation
+
+SIWE ReCaps unlock integration of protocols and/or APIs for developers by reducing user friction, onchain state and increasing security by introducing informed consent and deterministic capability objects on top of Sign-In With Ethereum (EIP-4361).
+
+While SIWE focuses on authenticating the Ethereum account against the service (relying party or SIWE client) initiating the SIWE flow, there is no canonical way to interact with a third-party service (resource service) on behalf of the authenticated Ethereum account. For example, a relying party might want to interact with another service on behalf of the Ethereum account, for example a service that provides data storage for the Ethereum account. This specification introduces a mechanism, that allows the service (or more generally a delegee) to combine authentication and authorization of such while preserving security and optimizing UX.
```
[the remark this answers is not in the document]
Comment (chunningham) [reply EIPS/eip-5573.md]: that does make things clearer, thanks

File: EIPS/eip-5573.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (awoie) [reply EIPS/eip-5573.md]: SIWE + Recap = OpenID (ID tokens) + OAuth2 (Access Tokens with DPoP and scopes)

File: EIPS/eip-5573.md [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+- ReCap SIWE Extension
+- ReCap Capability
+   - ReCap URI Scheme
+   - ReCap Details Object Schema
+- ReCap Translation Algorithm
+- ReCap Verification
+
+### ReCap SIWE Extension
+
+A ReCap is an EIP-4361 message following a specific format that allows an Ethereum account to delegate a set of ReCap Capabilities to a delegee through informed consent. Each ReCap Capability MUST be represented by an entry in the `Resources` array of the SIWE message that MUST deterministically translate the ReCap Capability in human-readable form to the `statement` field in the SIWE message using the ReCap Translation Algorithm.
```
[the remark this answers is not in the document]
Comment (chunningham) [reply EIPS/eip-5573.md]: I think you're right, it reads better as well

File: EIPS/eip-5573.md [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+- urn:recap:eyJhdHQiOnsiaHR0cHM6Ly9leGFtcGxlLmNvbSI6eyJleGFtcGxlL2FwcGVuZCI6W10sImV4YW1wbGUvcmVhZCI6W10sIm90aGVyL2FjdGlvbiI6W119LCJteTpyZXNvdXJjZTp1cmkuMSI6eyJleGFtcGxlL2FwcGVuZCI6W10sImV4YW1wbGUvZGVsZXRlIjpbXX0sIm15OnJlc291cmNlOnVyaS4yIjp7ImV4YW1wbGUvYXBwZW5kIjpbXX0sIm15OnJlc291cmNlOnVyaS4zIjp7ImV4YW1wbGUvYXBwZW5kIjpbXX19LCJwcmYiOltdfQ
+```
+
+#### ReCap Capability
+
+A ReCap Capability is identified by their ReCap URI that resolves to a ReCap Details Object which defines the associated actions and optional target resources. The scope of each ReCap Capability is attenuated by common fields in the SWIE message as described in the previous chapter, e.g., `URI`, `Issued At`, `Expiration Time`, `Not Before`.
+
+##### ReCap URI Scheme
+
+A ReCap URI starts with `urn:recap:` followed by `:` and the base64url-encoded payload of the ReCap Details Object. Note, the term base64url is defined in RFC4648 - Base 64 Encoding with URL and Filename Safe Alphabet.
```
[the remark this answers is not in the document]
Comment (chunningham) [reply EIPS/eip-5573.md]: specified it as unpadded

[the remark this answers is not in the document]
Comment (chunningham) [reply EIPS/eip-5573.md]: added a references section

**Review by skgbafa — APPROVED** [review]

**Pull request merged**